### PR TITLE
Implement zoning districts layer and hook up to filters

### PR DESCRIPTION
**Pull request by dhochbaum-dcp**
Completes ~~a portion of~~ #13

It adds Zustand for state management.  It sets up booleans to store the state of the image toggles for Zoning District and Tax Lots.  It also implements the Zoning District toggle, so that you can toggle on/off all of the layers.  It adds sets for the Zoning district categories and classes.

~~Moving forward, I was planning on structuring the store using top level booleans + layer group toggle array + layer ids checkbox array, and then using that to implement toggles/checkboxes for layer groups and layer ids.  More discussion can be found in Discussion #26.~~

~~There is a typescript error on an import.  I was unable to figure out what the source of the problem is.~~

Known outstanding issues:
- Toggling Residential Districts on causes some manufacturing districts to appear on the map.  They appear as purple and are distributed through most/all of the classes.
- Turning off all of the zd class checkboxes under a given category should automatically deselect that category, but it doesn't
- Turning off all category Switches should automatically hide the zoning district layer, but it does not. However, this functionality does work when the user deselects all zd class checkboxes
- The labels for the zoning districts do appear but the logic to hide them at certain zoom levels is not implemented
- The `f` feature argument for the Deck accessor functions such as `getFillColor` and `getFilterValue` is currently typed to `any`. We need to figure out how to strongly type that argument and update that in the code.

### CODE REVIEW

**Comment by TylerMatteo** [comment]
@dhochbaum-dcp  can you edit the PR summary above describing these changes, why you made the decisions you made, etc?

**Comment by TylerMatteo** [comment]
@dhochbaum-dcp I pushed up my new commit. A couple changes to note that aren't covered by my code comments:
* I moved `allZoningDistrictsVisibility` out of `getFilterValue` by assigning it to the `visible` property on the layer
* I changed `updateTriggers` to only trigger updates to `getFilterValue` and only based on the visible category and class Sets' `size` property. This should improve performance by making sure those function only rerun when absolutely necessary. Using `wholeStore` here would probably incur some unnecessary rerenders by Deck. I initially tried just passing in `visibleZoningDistrictCategories` and `visibleZoningDistrictClasses` themselves without `.size` but this didn't seem to work. I'm pretty sure that's because Deck isn't doing a deep comparison and, in JS, things like Sets, objects, and arrays are "by reference", therefore adding or removing things to the Set doesn't actually count as a "change". For our purposes here, `.size` seems like a good thing to trigger changes.

File: src/App.tsx [hunk 1]
```diff
@@ -36,15 +40,85 @@ function App() {
...
+  const allZoningDistrictsVisibility = useStore(
+    (state) => state.allZoningDistrictsVisibility,
+  );
+  const visibleZoningDistrictCategories = useStore(
+    (state) => state.visibleZoningDistrictCategories,
+  );
+  const visibleZoningDistrictClasses = useStore(
+    (state) => state.visibleZoningDistrictClasses,
+  );
+  const wholeStore = useStore((state) => state);
```
Comment (dhochbaum-dcp) [inline src/App.tsx:52]: I think this can be removed, as you switched to using .size instead.

**Comment by dhochbaum-dcp** [comment]
> * I moved `allZoningDistrictsVisibility` out of `getFilterValue` by assigning it to the `visible` property on the layer

My understanding is that clicking the image (which we toggling allZoningDistrictsVisibility) should show all of the layers when toggled on, and show only the layers selected in the class&category filters when toggled off.  In essence, it's a "quick show all the zoning layers" button, not a "show the zoning layers that I have selected below only if this is on" button.  Your change makes it the latter of the two.

**Comment by TangoYankee** [comment]
> > * I moved `allZoningDistrictsVisibility` out of `getFilterValue` by assigning it to the `visible` property on the layer
> 
> My understanding is that clicking the image (which we toggling allZoningDistrictsVisibility) should show all of the layers when toggled on, and show only the layers selected in the class&category filters when toggled off. In essence, it's a "quick show all the zoning layers" button, not a "show the zoning layers that I have selected below only if this is on" button. Your change makes it the latter of the two.

Would it be possible to get design/product management to weigh in? If there is differing interpretation among engineering, I think it's good practice to defer to the design/product management

**Comment by dhochbaum-dcp** [comment]
> > > * I moved `allZoningDistrictsVisibility` out of `getFilterValue` by assigning it to the `visible` property on the layer
> > 
> > 
> > My understanding is that clicking the image (which we toggling allZoningDistrictsVisibility) should show all of the layers when toggled on, and show only the layers selected in the class&category filters when toggled off. In essence, it's a "quick show all the zoning layers" button, not a "show the zoning layers that I have selected below only if this is on" button. Your change makes it the latter of the two.
> 
> Would it be possible to get design/product management to weigh in? If there is differing interpretation among engineering, I think it's good practice to defer to the design/product management

Yes, but (a) I wrote this after they had left for the day; and (b) I'm pretty confident in my interpretation that the big button shouldn't by default have no visible effect.  Obviously, it's best to check with design to confirm.

**Comment by jmquindala** [comment]
> > > * I moved `allZoningDistrictsVisibility` out of `getFilterValue` by assigning it to the `visible` property on the layer
> > 
> > 
> > My understanding is that clicking the image (which we toggling allZoningDistrictsVisibility) should show all of the layers when toggled on, and show only the layers selected in the class&category filters when toggled off. In essence, it's a "quick show all the zoning layers" button, not a "show the zoning layers that I have selected below only if this is on" button. Your change makes it the latter of the two.
> 
> Would it be possible to get design/product management to weigh in? If there is differing interpretation among engineering, I think it's good practice to defer to the design/product management

Just to make sure we're talking about the same thing, in this instance:
layer -> Zoning Districts or Tax Lots elements 
filters -> toggles and checkboxes 

In the design, I had it so that if you were to select a layer, it would turn on all the filters associated to that Layer. If you update the filters in any way, this should not affect the selection state of the layer. If you were to turn off a layer, then the filters associated to that layer should disappear from the UI and map. 

When a user re-selects a layer with filters they already manipulated, it should remember those filters. 

Let me know if you need further clarification!

**Comment by jmquindala** [comment]
> @jmquindala Thanks! One more question (for now): should going from 1 to 0 checked boxes toggle that category off, and the first box checked in a category automatically toggle it on?

Going from 1 to 0 checked boxes (assuming that this is the last checked box for every toggle within that layer's filter) should deselect that layer completely and hide that layer's accordion from the UI. Since the accordion for that layer is hidden, the user will not have the opportunity to change the selected filters from 0 to 1.

**Comment by TylerMatteo** [comment]
@dhochbaum-dcp I pushed some more changes to fix some issues with the filter toggling logic and to update the `getFillColor` code to use the keys of `colorKey` to do a lookup instead of matching a regex against `district`. I added some more known issues to the PR summary to be addressed in subsequent PRs but I'm going to merge this PR now before it starts blocking other work.

**Pull request merged**